File: swift-bridge/Sources/ContactsBridge/Predicates.swift
import Contacts
import Foundation

enum CNRContactPredicatePayload: Codable {
  case matchingName(name: String)
  case matchingEmailAddress(emailAddress: String)
  case matchingPhoneNumber(phoneNumber: CNRPhoneNumberPayload)
  case withIdentifiers(identifiers: [String])
  case inGroupWithIdentifier(groupIdentifier: String)
  case inContainerWithIdentifier(containerIdentifier: String)

  private enum CodingKeys: String, CodingKey {
    case kind
    case name
    case emailAddress
    case phoneNumber
    case identifiers
    case groupIdentifier
    case containerIdentifier
  }

  private enum Kind: String, Codable {
    case matchingName
    case matchingEmailAddress
    case matchingPhoneNumber
    case withIdentifiers
    case inGroupWithIdentifier
    case inContainerWithIdentifier
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    switch try container.decode(Kind.self, forKey: .kind) {
    case .matchingName:
      self = .matchingName(name: try container.decode(String.self, forKey: .name))
    case .matchingEmailAddress:
      self = .matchingEmailAddress(
        emailAddress: try container.decode(String.self, forKey: .emailAddress))
    case .matchingPhoneNumber:
      self = .matchingPhoneNumber(
        phoneNumber: try container.decode(CNRPhoneNumberPayload.self, forKey: .phoneNumber))
    case .withIdentifiers:
      self = .withIdentifiers(
        identifiers: try container.decode([String].self, forKey: .identifiers))
    case .inGroupWithIdentifier:
      self = .inGroupWithIdentifier(
        groupIdentifier: try container.decode(String.self, forKey: .groupIdentifier))
    case .inContainerWithIdentifier:
      self = .inContainerWithIdentifier(
        containerIdentifier: try container.decode(String.self, forKey: .containerIdentifier))
    }
  }

  func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    switch self {
    case .matchingName(let name):
      try container.encode(Kind.matchingName, forKey: .kind)
      try container.encode(name, forKey: .name)
    case .matchingEmailAddress(let emailAddress):
      try container.encode(Kind.matchingEmailAddress, forKey: .kind)
      try container.encode(emailAddress, forKey: .emailAddress)
    case .matchingPhoneNumber(let phoneNumber):
      try container.encode(Kind.matchingPhoneNumber, forKey: .kind)
      try container.encode(phoneNumber, forKey: .phoneNumber)
    case .withIdentifiers(let identifiers):
      try container.encode(Kind.withIdentifiers, forKey: .kind)
      try container.encode(identifiers, forKey: .identifiers)
    case .inGroupWithIdentifier(let groupIdentifier):
      try container.encode(Kind.inGroupWithIdentifier, forKey: .kind)
      try container.encode(groupIdentifier, forKey: .groupIdentifier)
    case .inContainerWithIdentifier(let containerIdentifier):
      try container.encode(Kind.inContainerWithIdentifier, forKey: .kind)
      try container.encode(containerIdentifier, forKey: .containerIdentifier)
    }
  }
}

enum CNRGroupPredicatePayload: Codable {
  case withIdentifiers(identifiers: [String])
  case subgroupsInGroupWithIdentifier(parentGroupIdentifier: String)
  case groupsInContainerWithIdentifier(containerIdentifier: String)

  private enum CodingKeys: String, CodingKey {
    case kind
    case identifiers
    case parentGroupIdentifier
    case containerIdentifier
  }

  private enum Kind: String, Codable {
    case withIdentifiers
    case subgroupsInGroupWithIdentifier
    case groupsInContainerWithIdentifier
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    switch try container.decode(Kind.self, forKey: .kind) {
    case .withIdentifiers:
      self = .withIdentifiers(
        identifiers: try container.decode([String].self, forKey: .identifiers))
    case .subgroupsInGroupWithIdentifier:
      self = .subgroupsInGroupWithIdentifier(
        parentGroupIdentifier: try container.decode(String.self, forKey: .parentGroupIdentifier))
    case .groupsInContainerWithIdentifier:
      self = .groupsInContainerWithIdentifier(
        containerIdentifier: try container.decode(String.self, forKey: .containerIdentifier))
    }
  }

  func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    switch self {
    case .withIdentifiers(let identifiers):
      try container.encode(Kind.withIdentifiers, forKey: .kind)
      try container.encode(identifiers, forKey: .identifiers)
    case .subgroupsInGroupWithIdentifier(let parentGroupIdentifier):
      try container.encode(Kind.subgroupsInGroupWithIdentifier, forKey: .kind)
      try container.encode(parentGroupIdentifier, forKey: .parentGroupIdentifier)
    case .groupsInContainerWithIdentifier(let containerIdentifier):
      try container.encode(Kind.groupsInContainerWithIdentifier, forKey: .kind)
      try container.encode(containerIdentifier, forKey: .containerIdentifier)
    }
  }
}

enum CNRContainerPredicatePayload: Codable {
  case withIdentifiers(identifiers: [String])
  case containerOfContactWithIdentifier(contactIdentifier: String)
  case containerOfGroupWithIdentifier(groupIdentifier: String)

  private enum CodingKeys: String, CodingKey {
    case kind
    case identifiers
    case contactIdentifier
    case groupIdentifier
  }

  private enum Kind: String, Codable {
    case withIdentifiers
    case containerOfContactWithIdentifier
    case containerOfGroupWithIdentifier
  }

  init(from decoder: Decoder) throws {
    let container = try decoder.container(keyedBy: CodingKeys.self)
    switch try container.decode(Kind.self, forKey: .kind) {
    case .withIdentifiers:
      self = .withIdentifiers(
        identifiers: try container.decode([String].self, forKey: .identifiers))
    case .containerOfContactWithIdentifier:
      self = .containerOfContactWithIdentifier(
        contactIdentifier: try container.decode(String.self, forKey: .contactIdentifier))
    case .containerOfGroupWithIdentifier:
      self = .containerOfGroupWithIdentifier(
        groupIdentifier: try container.decode(String.self, forKey: .groupIdentifier))
    }
  }

  func encode(to encoder: Encoder) throws {
    var container = encoder.container(keyedBy: CodingKeys.self)
    switch self {
    case .withIdentifiers(let identifiers):
      try container.encode(Kind.withIdentifiers, forKey: .kind)
      try container.encode(identifiers, forKey: .identifiers)
    case .containerOfContactWithIdentifier(let contactIdentifier):
      try container.encode(Kind.containerOfContactWithIdentifier, forKey: .kind)
      try container.encode(contactIdentifier, forKey: .contactIdentifier)
    case .containerOfGroupWithIdentifier(let groupIdentifier):
      try container.encode(Kind.containerOfGroupWithIdentifier, forKey: .kind)
      try container.encode(groupIdentifier, forKey: .groupIdentifier)
    }
  }
}

func cnrNativePredicate(_ predicate: CNRContactPredicatePayload) -> NSPredicate {
  switch predicate {
  case .matchingName(let name):
    CNContact.predicateForContacts(matchingName: name)
  case .matchingEmailAddress(let emailAddress):
    CNContact.predicateForContacts(matchingEmailAddress: emailAddress)
  case .matchingPhoneNumber(let phoneNumber):
    CNContact.predicateForContacts(matching: cnrDecodePhoneNumber(phoneNumber))
  case .withIdentifiers(let identifiers):
    CNContact.predicateForContacts(withIdentifiers: identifiers)
  case .inGroupWithIdentifier(let groupIdentifier):
    CNContact.predicateForContactsInGroup(withIdentifier: groupIdentifier)
  case .inContainerWithIdentifier(let containerIdentifier):
    CNContact.predicateForContactsInContainer(withIdentifier: containerIdentifier)
  }
}

func cnrNativePredicate(_ predicate: CNRGroupPredicatePayload) -> NSPredicate {
  switch predicate {
  case .withIdentifiers(let identifiers):
    CNGroup.predicateForGroups(withIdentifiers: identifiers)
  case .subgroupsInGroupWithIdentifier(let parentGroupIdentifier):
    CNGroup.predicateForSubgroupsInGroup(withIdentifier: parentGroupIdentifier)
  case .groupsInContainerWithIdentifier(let containerIdentifier):
    CNGroup.predicateForGroupsInContainer(withIdentifier: containerIdentifier)
  }
}

func cnrNativePredicate(_ predicate: CNRContainerPredicatePayload) -> NSPredicate {
  switch predicate {
  case .withIdentifiers(let identifiers):
    CNContainer.predicateForContainers(withIdentifiers: identifiers)
  case .containerOfContactWithIdentifier(let contactIdentifier):
    CNContainer.predicateForContainerOfContact(withIdentifier: contactIdentifier)
  case .containerOfGroupWithIdentifier(let groupIdentifier):
    CNContainer.predicateForContainerOfGroup(withIdentifier: groupIdentifier)
  }
}

File: src/predicates.rs
//! Predicate builders for contacts, groups, and containers.

use serde::{Deserialize, Serialize};

use crate::properties::CNPhoneNumber;

/// Mirrors the factory predicates from `CNContact+Predicates.h`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum CNContactPredicate {
    /// Matches contacts by name.
    MatchingName {
        /// The name to match.
        name: String,
    },
    /// Matches contacts by email address.
    MatchingEmailAddress {
        /// The email address to match.
        email_address: String,
    },
    /// Matches contacts by phone number.
    MatchingPhoneNumber {
        /// The phone number to match.
        phone_number: CNPhoneNumber,
    },
    /// Matches contacts by identifier.
    WithIdentifiers {
        /// The contact identifiers to match.
        identifiers: Vec<String>,
    },
    /// Matches contacts in a group.
    InGroupWithIdentifier {
        /// The group identifier to match.
        group_identifier: String,
    },
    /// Matches contacts in a container.
    InContainerWithIdentifier {
        /// The container identifier to match.
        container_identifier: String,
    },
}

impl CNContactPredicate {
    /// Creates a predicate matching the name.
    pub fn matching_name(name: impl Into<String>) -> Self {
        Self::MatchingName { name: name.into() }
    }

    /// Creates a predicate matching the email address.
    pub fn matching_email_address(email_address: impl Into<String>) -> Self {
        Self::MatchingEmailAddress {
            email_address: email_address.into(),
        }
    }

    /// Creates a predicate matching the phone number.
    pub fn matching_phone_number(phone_number: impl Into<CNPhoneNumber>) -> Self {
        Self::MatchingPhoneNumber {
            phone_number: phone_number.into(),
        }
    }

    /// Sets the identifiers.
    pub fn with_identifiers(identifiers: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self::WithIdentifiers {
            identifiers: identifiers.into_iter().map(Into::into).collect(),
        }
    }

    /// Creates a predicate for contacts in the group.
    pub fn in_group(group_identifier: impl Into<String>) -> Self {
        Self::InGroupWithIdentifier {
            group_identifier: group_identifier.into(),
        }
    }

    /// Creates a predicate for contacts in the container.
    pub fn in_container(container_identifier: impl Into<String>) -> Self {
        Self::InContainerWithIdentifier {
            container_identifier: container_identifier.into(),
        }
    }
}

/// Mirrors the factory predicates from `CNGroup+Predicates.h`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum CNGroupPredicate {
    /// Matches groups by identifier.
    WithIdentifiers {
        /// The group identifiers to match.
        identifiers: Vec<String>,
    },
    /// Matches subgroups within a group.
    SubgroupsInGroupWithIdentifier {
        /// The parent group identifier to match.
        parent_group_identifier: String,
    },
    /// Matches groups in a container.
    GroupsInContainerWithIdentifier {
        /// The container identifier to match.
        container_identifier: String,
    },
}

impl CNGroupPredicate {
    /// Sets the identifiers.
    pub fn with_identifiers(identifiers: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self::WithIdentifiers {
            identifiers: identifiers.into_iter().map(Into::into).collect(),
        }
    }

    /// Creates a predicate for subgroups in the group.
    pub fn subgroups_in_group(parent_group_identifier: impl Into<String>) -> Self {
        Self::SubgroupsInGroupWithIdentifier {
            parent_group_identifier: parent_group_identifier.into(),
        }
    }

    /// Creates a predicate for groups in the container.
    pub fn groups_in_container(container_identifier: impl Into<String>) -> Self {
        Self::GroupsInContainerWithIdentifier {
            container_identifier: container_identifier.into(),
        }
    }
}

/// Mirrors the factory predicates from `CNContainer+Predicates.h`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum CNContainerPredicate {
    /// Matches containers by identifier.
    WithIdentifiers {
        /// The container identifiers to match.
        identifiers: Vec<String>,
    },
    /// Matches the container for a contact.
    ContainerOfContactWithIdentifier {
        /// The contact identifier to match.
        contact_identifier: String,
    },
    /// Matches the container for a group.
    ContainerOfGroupWithIdentifier {
        /// The group identifier to match.
        group_identifier: String,
    },
}

impl CNContainerPredicate {
    /// Sets the identifiers.
    pub fn with_identifiers(identifiers: impl IntoIterator<Item = impl Into<String>>) -> Self {
        Self::WithIdentifiers {
            identifiers: identifiers.into_iter().map(Into::into).collect(),
        }
    }

    /// Creates a predicate for the container of the contact.
    pub fn container_of_contact(contact_identifier: impl Into<String>) -> Self {
        Self::ContainerOfContactWithIdentifier {
            contact_identifier: contact_identifier.into(),
        }
    }

    /// Creates a predicate for the container of the group.
    pub fn container_of_group(group_identifier: impl Into<String>) -> Self {
        Self::ContainerOfGroupWithIdentifier {
            group_identifier: group_identifier.into(),
        }
    }
}
